src/frontend/markdown.rs

```diff
@@ -181,6 +187,189 @@ impl InlineRule for MathEquationScanner {
+
+            entry.children.print_entry(fmt);
+
+            fmt.close("li");
+        });
+
+        fmt.close("ul");
+    }
+}
+
+impl NodeValue for Toc {
+    fn render(&self, node: &Node, fmt: &mut dyn Renderer) {
+        let mut attrs = node.attrs.clone();
+
+        attrs.push((
+            "class",
+            "not-prose float-left mr-20 w-80 menu rounded-box".into(),
+        ));
+
+        fmt.open("div", &attrs);
+
+        fmt.open("li", &[("class", "menu-title".into())]);
+        fmt.text("Table of Contents");
+        fmt.close("li");
+
+        self.print_entry(fmt);
+
+        fmt.close("div");
+    }
+}
+
+#[derive(Debug, Clone)]
+struct TocEntry {
+    level: u8,
+    sec_number: String,
+    name: String,
+    id: String,
+    children: Toc,
+}
+
+#[derive(Default)]
+struct TocBuilder {
+    top_level: Toc,
+    chain: Vec<TocEntry>,
+}
+
+impl TocBuilder {
+    fn new() -> TocBuilder {
+        TocBuilder::default()
+    }
+
+    fn into_toc(mut self) -> Toc {
+        self.fold_until(0);
+        self.top_level
+    }
+
+    fn fold_until(&mut self, level: u8) {
+        let mut this = None;
+        loop {
+            match self.chain.pop() {
+                Some(mut next) => {
+                    next.children.entries.extend(this);
+                    if next.level < level {
+                        self.chain.push(next);
+                        return;
+                    } else {
+                        this = Some(next);
+                    }
+                }
+                None => {
+                    self.top_level.entries.extend(this);
+                    return;
+                }
+            }
+        }
+    }
+
+    fn push(&mut self, level: u8, name: String, id: String) -> &str {
+        debug_assert!(level >= 1);
+
+        self.fold_until(level);
+
+        let mut sec_number;
+        {
+            let toc = match self.chain.last() {
+                None => {
+                    sec_number = String::new();
+                    &self.top_level
+                }
+                Some(entry) => {
+                    sec_number = entry.sec_number.clone();
+                    sec_number.push('.');
+                    &entry.children
+                }
+            };
+            let number = toc.count_entries_with_level(level);
+            sec_number.push_str(&(number + 1).to_string())
+        }
+
+        self.chain.push(TocEntry {
+            level,
+            name,
+            sec_number,
+            id,
+            children: Toc {
+                entries: Vec::new(),
+            },
+        });
+
+        let just_inserted = self.chain.last_mut().unwrap();
+        &just_inserted.sec_number
+    }
+}
+
+struct TocScanner;
+
+impl CoreRule for TocScanner {
+    fn run(root: &mut Node, _: &MarkdownIt) {
+        let mut slugger = Slugger::default();
+        let mut toc_builder = TocBuilder::new();
+        root.walk(|node, _| {
+            if node.is::<ATXHeading>() {
```

Comment:
There is also [SetExtheader](https://spec.commonmark.org/0.30/#setext-headings) which doesnt currently generate a toc entry.